### ISSUE TITLE
`refAutoReset` returns shallow ref

### ISSUE THREAD

**Issue by enkot** [vueuse/vueuse#3839]
### Describe the bug

Because `refAutoReset` uses `customRef` under the hood, it doesn't track deep mutations, i.e. it behaves as shallow ref, which is not expected.

I think it should return ref by default with additional 'shallow: boolean' option.

### Reproduction

https://stackblitz.com/edit/vitejs-vite-vk95hw?file=src%2FApp.vue

### System Info

```Shell
System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 18.18.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.2.3 - /usr/local/bin/npm
    pnpm: 8.15.3 - /usr/local/bin/pnpm
  npmPackages:
    @vueuse/core: ^10.9.0 => 10.9.0 
    vue: ^3.4.21 => 3.4.21
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vueuse/vueuse/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vueuse.org/guide).
- [X] Check that there isn't [already an issue](https://github.com/vueuse/vueuse/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a VueUse issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/core instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vueuse/vueuse/discussions).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by huangjw1997** [comment]
![image](https://github.com/vueuse/vueuse/assets/159113473/eb51f898-87d3-423a-a81e-71d951c38f3a)
源码出于设计确实只返回了对象的浅代理 你可以使用这种形式来解决问题
The source code by design does only return a shallow ref for the object, which you can use to solve the problem
```
 function changeName() {
  foo.value= {name:  'Ironman'} //  trigger rerendering
}
```

**Comment by OrbisK** [comment]
> it behaves as shallow ref, which is not expected.

Acording to this: https://github.com/vuejs/core/issues/10293 it is actually expected like this. I am not sure why the return type is not `ShallowRef` instead of `Ref`

**Comment by ilyaliao** [comment]
> > it behaves as shallow ref, which is not expected.它的行為與 淺 ref 類似，這是意料之外的。
> 
> According to this: [vuejs/core#10293](https://github.com/vuejs/core/issues/10293) it is actually expected like this. I am not sure why the return type is not `ShallowRef` instead of `Ref`根據這個： [vuejs/core#10293](https://github.com/vuejs/core/issues/10293) 實際上是這樣的預期。我不確定為什麼返回類型不是 `ShallowRef` 而是 `Ref`

I found that when the value changes, there is no log.

https://play.vuejs.org/#eNp9U8tunDAU/RXLG0CioOm0mykzUh9ZtFIfSquuLEWI3GFIjI38YBgh/j3XZnhESuINcO7xuec+6OnnpklaC3RHM12oqjFEg7HNgQlS1Y1UhvREQV6YqgUykKOSNQnwQvCJiZlgNfy83MJxJiTpBCUPOkAtDobUl2+5yd+T/cwPJ+Ww73b9ZbcZhihC4SwdvaAL/DBQNzw34DyR7LQ59P2klXTJhQxDliLqo5VorCHtu1reA98zuuIxSlIkZelKj8bU6EKKY1WiUSmwDb3TYbSQdVNxUL8bU0mhGd0RH3GxnHN5/uExoyzEE16coHh8AX/QncMY/aNAg2qB0TlmclWCGcM3f39Bh+9zEKuwHNlvBG9BS26dx5H2xYp7tL3iebff/agqUf7TN50BoaeinFHHHDyfURzt1zdKX+xukw/+HhMDdnE1buzhvBiF1UbWiMfknJvi9GyD3HCh88yjFbgHUiyb0ebcQjQmVriRSixiYWhUXjzGaL4qS1AR2R8mi1fu9YsQdBteZcbjr4bRAuD4sYWQcFmGAdKDmIzJF8pV1MMTOjbMHfxhQgHn/yvH4/F8XPcp+EpOvI85Z4W1VV/eYhabPT8GRF3n71pQbprY9m3yMdls6fAE0vQ9BA==

**Comment by OrbisK** [comment]
> > > it behaves as shallow ref, which is not expected.
> > 
> > 
> > According to this: [vuejs/core#10293](https://github.com/vuejs/core/issues/10293) it is actually expected like this. I am not sure why the return type is not `ShallowRef` instead of `Ref`
> 
> I found that when the value changes, there is no log in `set` scope.
> 
> https://play.vuejs.org/#eNp9U8tunDAU/RXLG0CioOm0mykzUh9ZtFIfSquuLEWI3GFIjI38YBgh/j3XZnhESuINcO7xuec+6OnnpklaC3RHM12oqjFEg7HNgQlS1Y1UhvREQV6YqgUykKOSNQnwQvCJiZlgNfy83MJxJiTpBCUPOkAtDobUl2+5yd+T/cwPJ+Ww73b9ZbcZhihC4SwdvaAL/DBQNzw34DyR7LQ59P2klXTJhQxDliLqo5VorCHtu1reA98zuuIxSlIkZelKj8bU6EKKY1WiUSmwDb3TYbSQdVNxUL8bU0mhGd0RH3GxnHN5/uExoyzEE16coHh8AX/QncMY/aNAg2qB0TlmclWCGcM3f39Bh+9zEKuwHNlvBG9BS26dx5H2xYp7tL3iebff/agqUf7TN50BoaeinFHHHDyfURzt1zdKX+xukw/+HhMDdnE1buzhvBiF1UbWiMfknJvi9GyD3HCh88yjFbgHUiyb0ebcQjQmVriRSixiYWhUXjzGaL4qS1AR2R8mi1fu9YsQdBteZcbjr4bRAuD4sYWQcFmGAdKDmIzJF8pV1MMTOjbMHfxhQgHn/yvH4/F8XPcp+EpOvI85Z4W1VV/eYhabPT8GRF3n71pQbprY9m3yMdls6fAE0vQ9BA==

yes. that is expected. because set will only trigger when writing to `.value`

**Comment by ilyaliao** [comment]
> > > > it behaves as shallow ref, which is not expected.
> > > 
> > > 
> > > According to this: [vuejs/core#10293](https://github.com/vuejs/core/issues/10293) it is actually expected like this. I am not sure why the return type is not `ShallowRef` instead of `Ref`
> > 
> > 
> > I found that when the value changes, there is no log in `set` scope.
> > [play.vuejs.org#eNp9U8tunDAU/RXLG0CioOm0mykzUh9ZtFIfSquuLEWI3GFIjI38YBgh/j3XZnhESuINcO7xuec+6OnnpklaC3RHM12oqjFEg7HNgQlS1Y1UhvREQV6YqgUykKOSNQnwQvCJiZlgNfy83MJxJiTpBCUPOkAtDobUl2+5yd+T/cwPJ+Ww73b9ZbcZhihC4SwdvaAL/DBQNzw34DyR7LQ59P2klXTJhQxDliLqo5VorCHtu1reA98zuuIxSlIkZelKj8bU6EKKY1WiUSmwDb3TYbSQdVNxUL8bU0mhGd0RH3GxnHN5/uExoyzEE16coHh8AX/QncMY/aNAg2qB0TlmclWCGcM3f39Bh+9zEKuwHNlvBG9BS26dx5H2xYp7tL3iebff/agqUf7TN50BoaeinFHHHDyfURzt1zdKX+xukw/+HhMDdnE1buzhvBiF1UbWiMfknJvi9GyD3HCh88yjFbgHUiyb0ebcQjQmVriRSixiYWhUXjzGaL4qS1AR2R8mi1fu9YsQdBteZcbjr4bRAuD4sYWQcFmGAdKDmIzJF8pV1MMTOjbMHfxhQgHn/yvH4/F8XPcp+EpOvI85Z4W1VV/eYhabPT8GRF3n71pQbprY9m3yMdls6fAE0vQ9BA==](https://play.vuejs.org/#eNp9U8tunDAU/RXLG0CioOm0mykzUh9ZtFIfSquuLEWI3GFIjI38YBgh/j3XZnhESuINcO7xuec+6OnnpklaC3RHM12oqjFEg7HNgQlS1Y1UhvREQV6YqgUykKOSNQnwQvCJiZlgNfy83MJxJiTpBCUPOkAtDobUl2+5yd+T/cwPJ+Ww73b9ZbcZhihC4SwdvaAL/DBQNzw34DyR7LQ59P2klXTJhQxDliLqo5VorCHtu1reA98zuuIxSlIkZelKj8bU6EKKY1WiUSmwDb3TYbSQdVNxUL8bU0mhGd0RH3GxnHN5/uExoyzEE16coHh8AX/QncMY/aNAg2qB0TlmclWCGcM3f39Bh+9zEKuwHNlvBG9BS26dx5H2xYp7tL3iebff/agqUf7TN50BoaeinFHHHDyfURzt1zdKX+xukw/+HhMDdnE1buzhvBiF1UbWiMfknJvi9GyD3HCh88yjFbgHUiyb0ebcQjQmVriRSixiYWhUXjzGaL4qS1AR2R8mi1fu9YsQdBteZcbjr4bRAuD4sYWQcFmGAdKDmIzJF8pV1MMTOjbMHfxhQgHn/yvH4/F8XPcp+EpOvI85Z4W1VV/eYhabPT8GRF3n71pQbprY9m3yMdls6fAE0vQ9BA==)
> 
> yes. that is expected. because set will only trigger when writing to `.value`

Yes, I'm not sure what else we can do to solve this issue. It seems we can't fix it in the short term as it's related to upstream dependencies.

**Comment by isimehmeti** [comment]
here is a workaround


https://stackblitz.com/edit/vitejs-vite-w6wpo4md?file=src%2FApp.vue

**Comment by ilyaliao** [comment]
I will close this.  For now, it can be used temporarily via https://github.com/vueuse/vueuse/issues/3839#issuecomment-2640050010."

**Comment by ilyaliao** [comment]
我發現另一個 `workaround`，我想符合`vue` 的設計理念。

https://stackblitz.com/edit/vitejs-vite-mrazstsy?file=src%2FApp.vue

**Comment by phk422** [comment]
> I found another `workaround`, and I think it aligns better with Vue's design philosophy.
> 
> https://stackblitz.com/edit/vitejs-vite-mrazstsy?file=src%2FApp.vue

This workaround did not trigger the automatic reset?

**Comment by ilyaliao** [comment]
> > Replying to **@ilyaliao** above
> >  
> > I found another `workaround`, and I think it aligns better with Vue's design philosophy.
> > https://stackblitz.com/edit/vitejs-vite-mrazstsy?file=src%2FApp.vue
> 
> This workaround did not trigger the automatic reset?

Thanks for catching that. Yes, sorry about that - the previous workaround was problematic.

`triggerRef` only triggers the customRef's **getter**, not the **setter**, so it won't trigger the reset.

I've updated the demo in StackBlitz to fix this.
